pytorch_lightning/trainer/training_tricks.py

```diff
@@ -19,9 +20,30 @@ def get_model(self):
         pass
 
     def clip_gradients(self):
+        # this code is a modification of torch.nn.utils.clip_grad_norm_
+        # with TPU support based on https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md
         if self.gradient_clip_val > 0:
             model = self.get_model()
-            torch.nn.utils.clip_grad_norm_(model.parameters(), self.gradient_clip_val)
+            parameters = model.parameters()
+            max_norm = self.gradient_clip_val
+            norm_type = 2
+            if isinstance(parameters, torch.Tensor):
+                parameters = [parameters]
+            parameters = list(filter(lambda p: p.grad is not None, parameters))
+            max_norm = float(max_norm)
+            norm_type = float(norm_type)
+            if norm_type == math.inf:
+                total_norm = max(p.grad.data.abs().max() for p in parameters)
+            else:
+                device = parameters[0].device
+                total_norm = torch.zeros([], device=device if parameters else None)
+                for p in parameters:
+                    param_norm = p.grad.data.norm(norm_type) ** norm_type
+                total_norm.add_(param_norm)
+                total_norm = (total_norm ** (1. / norm_type))
+            clip_coef = torch.tensor(max_norm, device=device) / (total_norm + 1e-6)
```

Comment:
this `1e-6` is kind of epsilon to avoid divide by 0, right?
what about `sys.float_info.epsilon`, see https://stackoverflow.com/questions/9528421/value-for-epsilon-in-python

Reply:
Is it okay if I keep this? It comes directly from the pytorch code which I am trying to stay close to.

Reply:
it is fine, I would just declare it in the module top like `EPSILON = 1e-6` so it is clear

Reply:
it is fine, I would just declare it in the module top like `EPSILON = 1e-6` so it is clear

Reply:
it is fine, I would just declare it in the module top like `EPSILON = 1e-6` so it is clear

Reply:
isn’t 1e-6 zero when on 16 bit? can we make it 1e-5 when 16 bit is on?

Reply:
Changed to 1e-6. Not sure what you are supposed to do with 16 bit (pytorch doesn't change the call)

Reply:
there is an option to check if it runs fp16 training `use_amp=hparams.use_16bit`

Reply:
Should I do this or just make eps 1e-5?

Reply:
Okay I changed it to check self.precisions